scipy/conftest.py

```diff
@@ -146,62 +148,76 @@ def check_fpu_mode(request):
-                    if 'cpu' not in torch.empty(0).device.type:
-                        pytest.skip(reason=reason)
-            return func(self, *args, **kwargs)
-    else:
-        @wraps(func)
-        def wrapped(*args, **kwargs):
-            xp = kwargs["xp"]
-            if SCIPY_ARRAY_API and SCIPY_DEVICE != 'cpu':
-                if xp.__name__ == 'cupy':
-                    pytest.skip(reason=reason)
-                elif xp.__name__ == 'torch':
-                    if 'cpu' not in torch.empty(0).device.type:
-                        pytest.skip(reason=reason)
-            return func(*args, **kwargs)
-    return wrapped
-
-
-def skip_if_array_api_backend(backend):
-    def wrapper(func):
-        reason = (
-            f"do not run with Array API backend: {backend}"
-        )
-        # method gets there as a function so we cannot use inspect.ismethod
-        if '.' in func.__qualname__:
-            @wraps(func)
-            def wrapped(self, *args, **kwargs):
-                xp = kwargs["xp"]
-                if xp.__name__ == backend:
-                    pytest.skip(reason=reason)
-                return func(self, *args, **kwargs)
-        else:
-            @wraps(func)
-            def wrapped(*args, **kwargs):
-                xp = kwargs["xp"]
-                if xp.__name__ == backend:
+@pytest.fixture
+def skip_if_array_api(xp, request):
+    """
+    Skip based on the ``skip_if_array_api`` marker.
+
+    Parameters
+    ----------
+    *backends : tuple
+        Backends to skip, e.g. ``("numpy.array_api", "torch")``.
+        These are overriden when ``np_only`` is ``True``, and are not
+        necessary to provide for non-CPU backends when ``cpu_only`` is ``True``.
+    reasons : list, optional
+        A list of reasons for each skip. When ``np_only`` is ``True``,
+        this should be a singleton list. Otherwise, this should be a list
+        of reasons, one for each corresponding backend in ``backends`.
+        If unprovided, default reasons are used. Note that it is not possible
+        to specify a custom reason with ``cpu_only``. Default: ``None``.
+    np_only : bool, optional
+        When ``True``, the test is skipped for all backends other
+        than the default NumPy backend. There is no need to provide
+        any ``backends`` in this case. To specify a reason, pass a
+        singleton list to ``reasons``. Default: ``False``.
+    cpu_only : bool, optional
+        When ``True``, the test is skipped on non-CPU devices.
+        There is no need to provide any ``backends`` in this case,
+        but any ``backends`` will also be skipped on the CPU.
+        Default: ``False``.
+    """
+    if "skip_if_array_api" not in request.keywords:
+        return
+    backends = request.keywords["skip_if_array_api"].args
+    kwargs = request.keywords["skip_if_array_api"].kwargs
+    try:
+        np_only = kwargs["np_only"]
+    except KeyError:
+        np_only = False
```

Comment:
NIT: `dict.get` is a bit more readable. Here and throughout this function

```suggestion
    np_only = kwargs.get("np_only", False)
```